gestalt-di/src/main/java/org/terasology/gestalt/di/BeanEnvironment.java

```diff
@@ -0,0 +1,314 @@
+            SoftServiceLoader<BeanDefinition> definitions = new SoftServiceLoader<>(BeanDefinition.class, ld, loadsFromParent);
+            ClassLookup lookup = new ClassLookup();
+
+            List<ClassRef<?>> namespaceIndex = new ArrayList<>();
+            Map<Class<?>, BeanDefinition<?>> definitionLookup = Maps.newHashMap();
+            Multimap<Class<?>, ClassRef> interfaceIndex = HashMultimap.create();
+            for (BeanDefinition<?> definition : definitions) {
+                // exclude objects that are from a different classloader
+                if (definition.targetClass().getClassLoader() != loader) {
+                    continue;
+                }
+                namespaceIndex.add(new ClassRef<>(definition));
+                definitionLookup.put(definition.targetClass(), definition);
+                for (Class<?> clazzInterface : definition.targetClass().getInterfaces()) {
+                    interfaceIndex.put(clazzInterface, new ClassRef<>(definition));
+                }
+            }
+            namespaceIndex.sort((Comparator<ClassRef>) (a1, a2) -> a1.prefix.compareTo(a2.prefix));
+            lookup.namespaceIndex = namespaceIndex.toArray(new ClassRef[0]);
+            lookup.interfaceIndex = interfaceIndex.asMap().entrySet().stream().collect(
+                    Collectors.toMap(k -> k.getKey(), v -> {
+                        ClassRef[] result = v.getValue().toArray(new ClassRef[0]);
+                        Arrays.sort(result, (classRef, t1) -> classRef.prefix.compareTo(t1.prefix));
+                        return result;
+                    })
+            );
+            lookup.definitions = definitionLookup;
+
+            return lookup;
+        });
+    }
+
+    public Iterable<ClassLoader> definitions() {
+        return beanLookup.keySet();
+    }
+
+    public boolean containsClass(Class<?> clazz) {
+        final ClassLookup lookup = beanLookup.get(clazz.getClassLoader());
+        return lookup.definitions.containsKey(clazz);
+    }
+
+    private Optional<Range<Integer>> findPrefixBounds(String prefix, ClassRef<?>[] input) {
+        if (input.length == 0) {
+            return Optional.empty();
+        }
+
+        if (Strings.isNullOrEmpty(prefix)) {
+            return Optional.of(Range.closed(0, input.length - 1));
+        }
+
+        int startPoint = 0;
+        int endpoint = input.length - 1;
+        while (true) {
+            String stringToTest = input[startPoint + (endpoint - startPoint) / 2].prefix;
+            if (stringToTest.startsWith(prefix)) {
+                break;
+            }
+            if (startPoint == endpoint) {
+                return Optional.empty();
+            }
+            if (stringToTest.compareTo(prefix) > 0) {
+                endpoint = startPoint + ((endpoint - startPoint) / 2);
+            }
+            if (stringToTest.compareTo(prefix) < 0) {
+                if (input[endpoint].prefix.compareTo(prefix) < 0) {
```

Comment:
I detect that this code is problematic. According to the [Correctness (CORRECTNESS)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#correctness-correctness), [SA: Self comparison of value with itself (SA_LOCAL_SELF_COMPARISON)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#sa-self-comparison-of-value-with-itself-sa-local-self-comparison).
This method compares a local variable with itself, and may indicate a typo or a logic error. Make sure that you are comparing the right things.

Reply:
input[endpoint].prefix and prefix are different variables

Reply:
Noticed :) thank you for your kind reply